### PR TITLE
mimalloc allocator

### PR DESCRIPTION
**Pull request by kile0**
**Description**: Add mimalloc as an alternate memory allocator. This avoids the dll-loading issue by linking mimalloc statically and calling the mi_malloc/mi_free apis explicitly.

**Motivation and Context**
- Why is this change required? What problem does it solve? Provides an alternate solution for models whose perf is adversely impacted by the BFC/default windows memory allocator. These are normally very lightweight models where the normal cost of memory allocation finally becomes significant compared to everything else.
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by snnn** [comment]
Hi @kile0,

Thanks for you contribution. Please temporarily set onnxruntime_USE_MIMALLOC to ON so that I can schedule a build and test your change.

File: cmake/onnxruntime_common.cmake [hunk 1]
```diff
@@ -48,10 +48,14 @@ source_group(TREE ${REPO_ROOT} FILES ${onnxruntime_common_src})
 add_library(onnxruntime_common ${onnxruntime_common_src})
 
 if (onnxruntime_USE_MIMALLOC)
-    include(external/mimalloc.cmake)
-    list(APPEND onnxruntime_EXTERNAL_LIBRARIES mimalloc-static)
-    list(APPEND onnxruntime_EXTERNAL_DEPENDENCIES mimalloc-static)
-    target_link_libraries(onnxruntime_common mimalloc-static)
+    if(onnxruntime_USE_CUDA OR onnxruntime_USE_OPENVINO)
```
Comment (snnn) [inline cmake/onnxruntime_common.cmake]: But we also have mkldnn/ngraph/...

Comment (kile0) [reply cmake/onnxruntime_common.cmake]: Is there a more flexible way to check for the cpu target?

Comment (snnn) [reply cmake/onnxruntime_common.cmake]: I'm not aware of any.

Comment (kile0) [reply cmake/onnxruntime_common.cmake]: Testing appears to be fine for ngraph/mkldnn so I'm going to leave the flag on for those two...it looks like the windows CI builds with mkldnn so at the very least that one is required. The original motivation for turning this off on CUDA/OpenVINO were strange crashes.

**Comment by kile0** [comment]
@snnn is it possible to rerun the three failing legs? All three windows failures passed on the previous CI run, and the intervening change is only cmake related

File: onnxruntime/core/framework/tensor.cc [hunk 2]
```diff
@@ -24,8 +24,14 @@ Tensor::Tensor(MLDataType p_type, const TensorShape& shape, std::shared_ptr<IAll
   if (shape_size < 0 || static_cast<uint64_t>(shape_size) >= std::numeric_limits<size_t>::max())
     ORT_THROW("shape.Size() must >=0");
   void* p_data = nullptr;
-  if (shape_size > 0)
-    p_data = allocator->AllocArray(static_cast<size_t>(shape_size), p_type->Size());
+  if (shape_size > 0) {
+    size_t len = 0;
+    if (!allocator->CalcMemSizeForArray(static_cast<size_t>(shape_size), p_type->Size(), &len))
```
Comment (snnn) [inline onnxruntime/core/framework/tensor.cc]: If it failed, please throw.

File: tools/ci_build/build.py [hunk 3]
```diff
@@ -332,7 +332,8 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
                  "-Donnxruntime_USE_AUTOML=" + ("ON" if args.use_automl else "OFF"),
                  "-Donnxruntime_CUDA_HOME=" + (cuda_home if args.use_cuda else ""),
                  "-Donnxruntime_USE_JEMALLOC=" + ("ON" if args.use_jemalloc else "OFF"),
-                 "-Donnxruntime_USE_MIMALLOC=" + ("ON" if args.use_mimalloc else "OFF"),
+                 "-Donnxruntime_USE_MIMALLOC=ON",
```
Comment (snnn) [inline tools/ci_build/build.py]: LGTM. Please revert this line back and I'll sign off it.

Comment (kile0) [reply tools/ci_build/build.py]: Thanks for reviewing!

**Comment by snnn** [comment]
Thanks

**Comment by kile0** [comment]
@snnn would you mind rerunning the CI checks? It looks like there were some infra issues...thanks!

**Comment by snnn** [comment]
Hi @kile0,  Yes, all the PRs are blocked now.

**Pull request merged**